### PR TITLE
SolrClient::optimize(): zpp segfault fix + parse int macros #3

### PR DESCRIPTION
**Pull request by omars44**
Fixes optimize segfault in php with debug mode.
optimize: Changed signature
internals: parse int macros (strict and non-strict modes)

### CODE REVIEW

File: src/php7/solr_macros.h [hunk 1]
```diff
@@ -170,6 +170,49 @@ static inline solr_ustream_t *solr_get_ustream_object(zend_object *obj)
...
+ * SOLR_PARSE_PARAM_INT(zval * zval_ptr, char * solrparam, size_t solrparam_len, const char * param_name, bool is_optional)
+ */
+#define SOLR_PARSE_PARAM_INT(zval_ptr, solrparam, solrparam_len, param_name) \
+    do { \
+        if (zval_ptr == NULL) { \
+            continue; \
+        } \
+        bool is_strict = ZEND_ARG_USES_STRICT_TYPES(); \
+        bool is_int = Z_TYPE_P(zval_ptr) == IS_LONG, is_str = Z_TYPE_P(zval_ptr) == IS_STRING; \
+        zend_uchar test = zval_get_type(zval_ptr); \
```
Comment (bix0r) [inline src/php7/solr_macros.h:198]: Unused `test` variable.
Also, could probably skip the `strict_on_and_not_int` variable and just do: `if (is_strict && !is_int) { \`

File: src/php7/solr_macros.h [hunk 2]
```diff
@@ -170,6 +170,49 @@ static inline solr_ustream_t *solr_get_ustream_object(zend_object *obj)
...
+ * Handles strict/non-strict integer parsing
+ *
+ * usage:
+ * 		SOLR_PARSE_PARAM_INT_DEF(maxSegmentsZval, maxSegments, maxSegmentsLen, "1");
+ * 		if (zend_parse_parameters(ZEND_NUM_ARGS(), "z", &maxSegmentsZval) == FAILURE) {
+ *			RETURN_THROWS();
+ *		}
+ *		SOLR_PARSE_PARAM_INT(maxSegmentsZval, maxSegments, maxSegmentsLen, "maxSegments");
+ */
+#define SOLR_PARSE_PARAM_INT_DEF(zval_ptr, solrparam, solrparam_len, default_value) \
```
Comment (bix0r) [inline src/php7/solr_macros.h:183]: I don't like the name of this macro, "DEF" sounds too much like "default". But I don't like `SOLR_PARSE_PARAM_INT_DEFINITION` either :confused: 

`SOLR_DEFINE_PARSE_PARAM_INT` ? (not much better)

File: src/php7/php_solr_client.c [hunk 3]
```diff
@@ -1564,13 +1564,12 @@ PHP_METHOD(SolrClient, optimize)
...
-	if (zend_parse_parameters(ZEND_NUM_ARGS(), "|sbb", &maxSegments, &maxSegmentsLen, &softCommit, &waitSearcher) == FAILURE) {
-
-		php_error_docref(NULL, E_WARNING, "Invalid parameter");
-
-		return;
+	if (zend_parse_parameters(ZEND_NUM_ARGS(), "|zbb", &maxSegmentsZval, &softCommit, &waitSearcher) == FAILURE) {
+		RETURN_THROWS();
 	}
 
+	SOLR_PARSE_PARAM_INT(maxSegmentsZval, maxSegments, maxSegmentsLen, "maxSegments");
```
Comment (bix0r) [inline src/php7/php_solr_client.c:1571]: I don't like to change the signature here this drastically, unless we want to bump the major version.
If someone is using `optimize()` today in strict mode, then that code has been forced to provide a string `$maxSegments`. With this change we are throwing an exception when the extension is updated.

I would prefer we actually just change the `SOLR_PARSE_PARAM_INT` macro to allow strings in strict mode for now, and when we want do bump the major version, we can just change the macro to be properly strict :)

**Review by bix0r — CHANGES_REQUESTED** [review]
I like the new macros, but I have concerns regarding the signature change.

File: src/php7/php_solr_client.c (hunk 3 above)
Comment (omars44) [reply src/php7/php_solr_client.c:1571]: sounds reasonable, I was just thinking about it

**Comment by omars44** [comment]
@bix0r Addressed all the comments

File: src/php7/solr_macros.h [hunk 4]
```diff
@@ -170,6 +170,50 @@ static inline solr_ustream_t *solr_get_ustream_object(zend_object *obj)
...
+ * SOLR_PARSE_PARAM_INT(zval * zval_ptr, char * solrparam, size_t solrparam_len, const char * param_name, bool is_optional)
+ */
+#define SOLR_PARSE_PARAM_INT(zval_ptr, solrparam, solrparam_len, param_name) \
+    do { \
+        if (zval_ptr == NULL) { \
+            continue; \
+        } \
+        bool is_strict = ZEND_ARG_USES_STRICT_TYPES(); \
+        bool is_int = Z_TYPE_P(zval_ptr) == IS_LONG, is_str = Z_TYPE_P(zval_ptr) == IS_STRING; \
+        bool strict_on_and_not_int = is_strict && Z_TYPE_P(zval_ptr) != IS_LONG; \
```
Comment (bix0r) [inline src/php7/solr_macros.h:198]: Can you remove this unused line here?

**Review by bix0r — CHANGES_REQUESTED** [review]
Looks good :+1: 
Just fix that unused line

**Review by bix0r — APPROVED** [review]

**Pull request merged**